adapter/internal/oasparser/envoyconf/routes_with_clusters.go

```diff
@@ -182,37 +184,11 @@ func CreateRoutesWithClusters(mgwSwagger model.MgwSwagger, upstreamCerts map[str
 		}
 	}
 
-	apiRequestInterceptor = mgwSwagger.GetInterceptor(mgwSwagger.GetVendorExtensions(), xWso2requestInterceptor, APILevelInterceptor)
-	// if lua filter exists on api level, add cluster
-	if apiRequestInterceptor.Enable {
-		logger.LoggerOasparser.Debugf("API level request interceptors found for %v : %v", apiTitle, apiVersion)
-		apiRequestInterceptor.ClusterName = getClusterName(requestInterceptClustersNamePrefix, organizationID, vHost,
-			apiTitle, apiVersion, "")
-		cluster, addresses, err := CreateLuaCluster(interceptorCerts, apiRequestInterceptor)
-		if err != nil {
-			apiRequestInterceptor = model.InterceptEndpoint{}
-			logger.LoggerOasparser.Errorf("Error while adding api level request intercepter external cluster for %s. %v",
-				apiTitle, err.Error())
-		} else {
-			clusters = append(clusters, cluster)
-			endpoints = append(endpoints, addresses...)
-		}
-	}
-	apiResponseInterceptor = mgwSwagger.GetInterceptor(mgwSwagger.GetVendorExtensions(), xWso2responseInterceptor, APILevelInterceptor)
-	// if lua filter exists on api level, add cluster
-	if apiResponseInterceptor.Enable {
-		logger.LoggerOasparser.Debugln("API level response interceptors found for " + mgwSwagger.GetID())
-		apiResponseInterceptor.ClusterName = getClusterName(responseInterceptClustersNamePrefix, organizationID, vHost,
-			apiTitle, apiVersion, "")
-		cluster, addresses, err := CreateLuaCluster(interceptorCerts, apiResponseInterceptor)
-		if err != nil {
-			apiResponseInterceptor = model.InterceptEndpoint{}
-			logger.LoggerOasparser.Errorf("Error while adding api level response intercepter external cluster for %s. %v", apiTitle, err.Error())
-		} else {
-			clusters = append(clusters, cluster)
-			endpoints = append(endpoints, addresses...)
-		}
-	}
+	// Create API level interceptor clusters if required
+	clustersI, endpointsI, apiRequestInterceptor, apiResponseInterceptor := createInterceptorAPIClusters(mgwSwagger,
+		interceptorCerts, vHost, organizationID)
+	clusters = append(clusters, clustersI...)
```

Comment:
what happens when clustersI == nil i.e. when there's no interceptors?

Reply:
btw, can't we move this entire block out of the CreateRoutesWithClusters to a new method and refer that new method from GetRoutesClustersEndpoints method.

Reply:
@AmaliMatharaarachchi 

1. `createInterceptorAPIClusters()` returns `[]*clusterv3.Cluster` and `[]*corev3.Address`. So if no intercepters that returned array of pointers will be empty and passed to the append method and nothing gets appended.

2. Yeah, I initially thought of moving the entire block out from `CreateRoutesWithClusters` but we pass request interceptors and response interceptors per resource to the `createRoute` method inside  `CreateRoutesWithClusters`. So if we are to completely take the interceptor creation code block out, we need to do following steps or something similar to that.

- Call `createInterceptorAPIClusters` from `GetRoutesClustersEndpoints` and save the interceptors in a map. We could probably have `resource:interceptor` mapping there.
- Pass the created interceptors to `CreateRoutesWithClusters` and in the for loop where we do `for _, resource := range mgwSwagger.GetResources()`, we can get the created interceptors using the resource key.

WDYT about that approach ?

Reply:
I suppose we are good with your above implementation then.

Reply:
@AmaliMatharaarachchi Another approach is to separate cluster creation and route creation. We can split the logic inside `CreateRoutesWithClusters` into two.

1.   Create production, sandbox clusters, xwso2 clusters, API request and response interceptors first and store them in variables / maps.  (We are creating these per API not per resource)
2. In the next step we can iterate the resources and create resource and operational level interceptors if required and then finally call `createRoute` to create the route. We need to pass the clusters and API interceptors created in 1 and use them here. 

In this approach also we'll have to store the resources created in 1 and use them in 2.

I'll update this PR once the operational policies PR gets merged